[ENH] Add Time Series Transformer Classifier

TenFinges opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at our
contribution guide: https://www.aeon-toolkit.org/en/latest/contributing.html.

Feel free to delete sections of this template if they do not apply to your PR,
avoid submitting a blank template or empty sections.
If you are a new contributor, do not delete this template without a suitable
replacement or reason. If in doubt, ask for help. We're here to help!

Please be aware that we are a team of volunteers so patience is
necessary when waiting for a review or reply. There may not be a quick turnaround for
reviews during slow periods. While we value all contributions big or small, pull
requests which do not follow our guidelines may be closed.
-->

#### Reference Issues/PRs

None

<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

#### What does this implement/fix? Explain your changes.

This PR adds a Transformer-based classifier for time series classification, filling a significant gap in the deep learning module.

It implements two new classes:

1. aeon.networks.TransformerNetwork: A pure Keras/TensorFlow implementation of the Transformer Encoder architecture (Input Embedding -> Positional Encoding -> Transformer Blocks -> GAP).
2. aeon.classification.deep_learning.TimeTransformerClassifier: A classifier wrapper that manages the network training and prediction, fully compatible with the aeon and scikit-learn API.

The implementation follows existing patterns in aeon (e.g., InceptionTime, ResNet) and includes unit tests to verify functionality.


<!--
A clear and concise description of what you have implemented.
-->

#### Does your contribution introduce a new dependency? If yes, which one?

No. It relies on tensorflow, which is already an existing soft dependency for the 'deep_learning' module.

<!--
If your contribution does add a dependency, we may suggest adding it as an
optional/soft dependency to keep external dependencies of the core aeon package
to a minimum.
-->

#### Any other comments?

Implementation includes comprehensive docstrings and unit tests (test_transformer.py) which verify the model construction and training loop.


<!--
Any other information that is important to this PR or helpful for reviewers.
-->

### PR checklist

<!--
Please go through the checklist below. Please feel free to remove points if they are
not applicable. To check a box, replace the space inside the square brackets with an
'x' i.e. [x].
-->

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you **after** the PR has been merged.
- [x] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

##### For new estimators and functions
- [x] I've added the estimator/function to the online [API documentation](https://www.aeon-toolkit.org/en/latest/api_reference.html).

<!--
Thanks for contributing!
-->

TenFinges committed:
[ENH] Add Time Series Transformer Classifier

TenFinges committed:
[DOC] Add Transformer Classifier and Network to API docs

TenFinges committed:
[BUG] Fix CI: Correct n_epochs param and allow list activation

TenFinges committed:
[BUG] Fix CI: Add save_init_model param and ensure deterministic random_state

TenFinges committed:
[BUG] Fix CI: Propagate last_file_name to BaseDeepClassifier

hadifawaz1999 reviewed (changes requested):
Thank you for the contribution, its been on my to do list to take care of this, we've been planning to port in this classifier [here](https://github.com/MSD-IRIMAS/DeepRehabPile/blob/main/deep_rehab_pile/classifiers/_van_tran.py) as its based on a publication as mentioned in its docstring, it would be great if you can adapt the network to be like this one

TenFinges committed:
[REF] Adapt TransformerNetwork to VanTran architecture with sinusoidal APE

TenFinges committed:
[FIX] Resolve Core Import, Keras Serialization, and Flake8 issues

1. Fix Core Import Failure: Use duplicate APE definition inside _get_ape_class helper in networks/_transformer.py to allow lazy TensorFlow imports. 2. Fix Keras Serialization: Register APE class with @tf.keras.utils.register_keras_serializable and overwrite load_model in TimeTransformerClassifier. 3. Fix NameError by ensuring numpy is imported in classifier. 4. Fix F401 unused import.

TenFinges committed:
Merge branch 'main' into feat/transformer-classifier

TonyBagnall committed:
Merge branch 'main' into feat/transformer-classifier
